### PR TITLE
Limit garbage cleanup speed to avoid blocking snapshot apply.

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

## What have you changed? (mandatory)
Limit garbage cleanup speed to avoid blocking snapshot apply. So when a TiKV re-join the cluster, it can still apply snapshots even if it has a lot of garbage to cleanup.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
unit tests.

Please describe the tests that you ran to verify your changes. Have you finished unit tests, integration tests, or manual tests? What additional tests would give you greater confidence in this change?

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Please fix compiling error.

File: src/raftstore/store/worker/region.rs [hunk 1]
```diff
@@ -48,6 +48,7 @@ const GENERATE_POOL_SIZE: usize = 2;
 
 // used to periodically check whether we should delete a stale peer's range in region runner
 pub const STALE_PEER_CHECK_INTERVAL: u64 = 10_000; // milliseconds
+const CLEANUP_MAX_BLOCK_DURATION: Duration = Duration::from_secs(5);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/region.rs]: S/CLEANUP_MAX_BLOCK_DURATION/CLEANUP_MAX_DURATION

File: src/raftstore/store/worker/region.rs [hunk 2]
```diff
@@ -416,27 +408,28 @@ impl SnapContext {
             .pending_delete_ranges
             .drain_overlap_ranges(start_key, end_key);
         for (region_id, s_key, e_key) in overlap_ranges {
-            self.cleanup_range(region_id, s_key, e_key, false /* use_delete_files */);
+            let use_delete_files = false;
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/region.rs]: Please move use_delete_files out of the loop.

File: src/raftstore/store/worker/region.rs [hunk 3]
```diff
@@ -448,11 +441,24 @@ impl SnapContext {
...
+                    use_delete_files,
+                );
+                cleaned_range_keys.push(start_key);
+                if now.elapsed() >= CLEANUP_MAX_BLOCK_DURATION {
+                    break;
+                }
+            }
+        }
+        for key in cleaned_range_keys {
+            self.pending_delete_ranges.remove(&key);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/region.rs]: I think remove should return the removed item, so here you can check if the removed item is expected.

File: src/raftstore/store/worker/region.rs [hunk 4]
```diff
@@ -162,14 +163,14 @@ impl PendingDeleteRanges {
         ranges
     }
 
+    fn remove(&mut self, start_key: &[u8]) -> Option<(u64, Vec<u8>, Vec<u8>)> {
```
Comment (siddontang) [inline src/raftstore/store/worker/region.rs]: why not return PeerInfo directly? strange to return a tuple here.

Comment (hicqu) [reply src/raftstore/store/worker/region.rs]: Because `insert` needs a tuple, so I think I should keep 2 interfaces similar.

**Comment by zhangjinpeng1987** [comment]
@MyonKeminta PTAL

File: src/raftstore/store/worker/region.rs (hunk 4 above)
Comment (MyonKeminta) [inline src/raftstore/store/worker/region.rs]: Is the return value needed?

Comment (hicqu) [reply src/raftstore/store/worker/region.rs]: I think so, it's used for `assert` something.

**Pull request merged**